### PR TITLE
add new adatptor doc

### PR DESCRIPTION
**Pull request by ClarkChin08**
## Type of Change

Doc add new backend

## Description

add new backend of how to add new adaptor

## How has this PR been tested?

Spelling typo

### CODE REVIEW

File: docs/source/add_new_adaptor.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+- [Introduction](#introduction)
+- [Add query_fw_capability to Adaptor](#add-query-fw-capability-to-adaptor)
+- [Add Pre-quantization Optimizations](#add-pre-quantization-optimizations)
+- [Quantize Model according to tune cfg](#quantize-model-according-to-tune-cfg)
+- [Evaluate Model on Validation Dataset](#evaluate-model-on-validation-dataset)
+- [Train Model for Quantization-aware Training](#Train-model-for-quantization-awar-training)
+- [Other API that not Mandatory Needed](#other-api-that-not-mandatory-needed)
+- [Summary](#summary)
+
+## Introduction
```
Comment (kevinintel) [inline docs/source/add_new_adaptor.md:16]: please point out why you should write thsi document.

File: docs/source/add_new_adaptor.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+|save          |used by tune strategy class for saving model                 |
+|convert       |convert a source model format to another                     |
+|quantize_input|quantize the model to be able to take quantized input        |
+|inspect_tensor|used by tune strategy class for dumping tensor info          |
+|set_tensor    |used by tune strategy class for setting tensor back to model |
+|convert_bf16  |execute bf16 model conversion                                |
+
+
+## Summary
+The document outlines the process of adding support for a new adaptor, in Intel® Neural Compressor with minimal changes. It provides instructions and code examples for implementation of a new backend. By following the steps outlined in the document, users can extend Intel® Neural Compressor's functionality to accommodate new adaptor and incorporate it into quantization workflows.
```
Comment (kevinintel) [inline docs/source/add_new_adaptor.md:243]: It's better to move to top

**Review by kevinintel — CHANGES_REQUESTED** [review]
please refine it

File: docs/source/add_new_adaptor.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
+
+
+How to Support New Adaptor
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:3]: 'support' to 'add an'

File: docs/source/add_new_adaptor.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+- [Train Model for Quantization-aware Training](#Train-model-for-quantization-awar-training)
+- [Other API that not Mandatory Needed](#other-api-that-not-mandatory-needed)
+- [Summary](#summary)
+
+## Introduction
+Intel® Neural Compressor builds the low-precision inference solution on popular deep learning frameworks such as TensorFlow, PyTorch, MXNet, Keras and ONNX Runtime. The adaptor layer is the bridge between the tuning strategy and vanilla framework quantization APIs.
+
+The quantizable operator behavior and it's tuning scope is defined in specific framework YAML file. The adaptor will parse this file and give the quantization capability to the Strategy object. Then Strategy will build tuning space of the specific graph/model and generate different tuning configuration from the tuning space to adaptor.
+
+The diagram below illustrates all the relevant steps of how adaptor is invoked, with additional details provided for each annotated step.
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:21]: we should introduce what's the interface in adaptor layer at first, and then the working flow how they interactive each other

File: docs/source/add_new_adaptor.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    Strategy ->> Adaptor: query framework capability
+    Adaptor ->> Strategy: Parse the framework YAML and return capability
+    Strategy ->> Strategy: Build tuning space
+    loop Traverse tuning space
+    	Strategy->> Adaptor: generate next tuning cfg
+        Adaptor ->> Adaptor: calibrate and quantize model based on tuning config
+
+    end
+```
+1. **Strategy**: Drives the overall tuning process and utilizes `adaptor.query_fw_capability` to query the framework's capabilities.
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:34]: there are a lots of `strategy` words here. the organization is NOT straightforward, at least for me.

File: docs/source/add_new_adaptor.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+After the framework YAML enabled, The The abstract class `QueryBackendCapability` should be inherited and implemented. This class is used to load the YAML file and parse the quantization config from the specific framework version. You can refer to Tensorflow implementation [TensorflowQuery](../neural_compressor/adaptor/tensorflow.py#L628).
+
+With the inherited `QueryBackendCapability` class enabled. The Adaptor will initialize a `query_handler` and used in API `query_fw_capability`.
+
+Each framework adaptor should implement the `query_fw_capability` fucntion, this function will only be invoked once and will loop over the graph/model for the quantizable operators and collect each operator's opwise details and optypewise capability. You should return a standard dict of the input model's tuning capability. The format is like below:
+
+```python
+    def query_fw_capability(self, model):
+        # loop over the model and generate the capability of the model
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:58]: this code snippet is also not easy to understand, such as where the quantiable_op_details variable comes from? whether it's the whole code or just partial of code? should use `...` to present those skipped code?

File: docs/source/add_new_adaptor.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+As the `query_fw_capability` only invoked once, we can do some pre-optimizations in the `query_fw_capability` API and cache the pre_optimized_object for quantization. It's not mandatory and if you have done optimizations on the model, take care to use the optimized model in the quantize API.
+
+
+## Quantize Model according to tune_cfg
+`quantize` function is used to perform quantization for post-training quantization and quantization-aware training. Quantization processing includes calibration and conversion processing for post-training quantization, while for quantization-aware training, it includes training and conversion processing.
+
+The first work of `quantize` function is to invoke `tuning_cfg_to_fw` to generate the self.quantize_config. The self.quantize_config is a dict including the quantization information. It's format is like below
+
+```
+self.quantize_config  = {
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:132]: maybe integrate with working flow is better? I think here is used to describe what's the input format of the quantize() func in adaptor, right?

File: docs/source/add_new_adaptor.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+              is_asymmetric, 
+              weight_bit)
+} 
+```
+
+You can also set bf16_ops in `tuning_cfg_to_fw` and the `self.bf16_ops` will be converted in `convert_bf16` function.
+
+After got the `self.quantize_config`, we can prepare to quantize the model. It usually have three steps.
+
+### Insert FakeQuant operator to the fp32 graph for calibration.
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:157]: this is keras specific implementation, maybe you can skip this?

**Comment by chensuyue** [comment]
Please clean the CI issue, add reasonable words into https://github.com/intel/neural-compressor/blob/add_new_backend/.azure-pipelines/scripts/codeScan/pyspelling/inc_dict.txt

**Comment by chensuyue** [comment]
@ClarkChin08 please take a look at the comments.

File: docs/source/add_new_adaptor.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    Adaptor ->> Adaptor: Pre-optimize the input fp32 model as self.optimized_model
+    Adaptor ->> Strategy: Send the capability including 'opwise' and 'optypewise' ability
+    Strategy ->> Strategy: Build tuning space
+    loop Traverse tuning space
+    	Strategy->> Adaptor: generate next tuning cfg
+        Adaptor ->> Adaptor: calibrate and quantize model based on tuning config
+
+    end
+```
+1. **Adaptor**: Design the framework YAML, inherit `QueryBackendCapability` class to parse the framework yaml.
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:41]: still like what I mentioned, this index + *Adaptor* is not a good way to describe what happens in the diagram. 

is it better to use ① like symbol and skip the first word like **adaptor**?

File: docs/source/add_new_adaptor.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+5. **Adaptor**: Pre-optimize the input fp32 model as self.optimized_model
+6. **Adaptor**: Send the capability including 'opwise' and 'optypewise' ability to Strategy
+7. **Strategy**: Generates the tuning configurations for each operators of the model using the tuning space constructed in the previous step, specifying the desired tuning process.
+8. **Adaptor**: Invokes the specific kernels for the calibration and quantization based on the tuning configuration.
+
+## API List that Need to Implement
+These APIs are necessary to add a new adapter. Here are the parameter types and functionality descriptions of these APIs. The following chapters will introduce the specific implementation and data format of these APIs in detail.
+| API | Parameters | Usage | Comments|
+| :------ | :------ | :------ | :------ |
+| query_fw_capability(self, model) | **model** (object): A INC model object to query quantization tuning capability. | The function is used to return framework tuning capability. |Confirm the the data format output by the function must meet the requirements |
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:54]: what's the output of such funcs? could you pls add?

File: docs/source/add_new_adaptor.md (hunk 6 above)
Comment (ftian1) [reply docs/source/add_new_adaptor.md:73]: you didn't solve my concerns.

I think the doc is too complex and expose too much internals to user like the explanation on `quantizable_op_details`

File: docs/source/add_new_adaptor.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+| API | Parameters |Output   |Usage   | Comments|
+| :------ | :------|:------ |:------ | :------ |
+| query_fw_capability(self, model) | **model** (object): A INC model object to query quantization tuning capability. |output format {'opwise': {(node_name, node_op): [{'weight': {'dtype': ...#int8/fp32 or other data type}, 'activation': {'dtype': ...#int8/fp32 or other data type}}, ...]}, 'optypewise':{node_op: [{'weight': {'dtype': ...#int8/fp32 or other data type}, 'activation': {'dtype': ...#int8/fp32}}], ...}} |The function is used to return framework tuning capability. |Confirm the the data format output by the function must meet the requirements |
+| get_optype_wise_ability(self, quantizable_op_details) | **quantizable_op_details** (string dict): the key is op type while the value is the detail configurations of activation and weight for this op type. |output optypewise capability with format {node_op: {'weight': {'dtype': ...#int8/fp32 or other datatype}, 'activation': {'dtype': ...#int8/fp32 or other datatype}}], ...} |Used in query_fw_capability. When get the quantizable_op_details, use this API to generate optype wise ability | |
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md]: shall we remove this line as it doesn't need to be aware by user?

File: docs/source/add_new_adaptor.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+| quantize(self, tune_cfg, model, dataloader, q_func=None) | **tune_cfg** (dict): the chosen tuning configuration.<br> **model** (object): The model to do quantization.<br>**dataloader** (object): The dataloader used to load quantization dataset. **q_func**(optional): training function for quantization aware training mode.| output quantized model object|This function use the dataloader to generate the data required by the model, and then insert Quantize/Dequantize operator into the quantizable op required in the tune_config and generate the model for calibration, after calibration, generate the final quantized model according to the obtained data range from calibration| |
+| tuning_cfg_to_fw(self, tuning_cfg) | **tuning_cfg** (string dict): Tuning config generated . |None| The function is used in quantize API and transfer the chosen tuning config to self.quantize_config.|Confirm the the data format output by the function must meet the requirements |
```
Comment (ftian1) [inline docs/source/add_new_adaptor.md:50]: it looks more like internal implementation but not the interface user needs to be aware of

**Review by ftian1 — APPROVED** [review]
most are ok to me. just minor comments
